atomicfu-gradle-plugin/src/main/kotlin/kotlinx/atomicfu/plugin/gradle/AtomicFUGradlePlugin.kt

```diff
@@ -254,50 +245,73 @@ private fun Project.configureTransformationForTarget(target: KotlinTarget) {
+                        originalClassesDirs,
+                        config
+                    )
+                    .also {
+                        compilation.defaultSourceSet.kotlin.compiledBy(it, AtomicFUTransformTask::destinationDirectory)
+                    }
             }
             KotlinPlatformType.js -> {
                 // skip when js transformation is not needed or when IR is transformed
                 if (!config.transformJs || (needsJsIrTransformation(target))) {
                     return@compilations
                 }
-                project.createJsTransformTask(compilation).configureJsTask(
-                    compilation.compileAllTaskName,
-                    transformedClassesDir,
-                    originalClassesDirs,
-                    config
-                )
+                project.registerJsTransformTask(compilation)
+                    .configureJsTask(
+                        compilation.compileAllTaskName,
+                        transformedClassesDir,
+                        originalClassesDirs,
+                        config
+                    )
+                    .also {
+                        compilation.defaultSourceSet.kotlin.compiledBy(it, AtomicFUTransformJsTask::destinationDirectory)
+                    }
             }
             else -> error("Unsupported transformation platform '${target.platformType}'")
         }
         //now transformTask is responsible for compiling this source set into the classes directory
+        compilation.defaultSourceSet.kotlin.destinationDirectory.value(transformedClassesDir)
         classesDirs.setFrom(transformedClassesDir)
-        classesDirs.builtBy(transformTask)
-        (tasks.findByName(target.artifactsTaskName) as? Jar)?.apply {
-            setupJarManifest(multiRelease = config.jvmVariant.toJvmVariant() == JvmVariant.BOTH)
+        classesDirs.setBuiltBy(listOf(transformTask))
+        tasks.withType(Jar::class.java).configureEach {
+            if (name == target.artifactsTaskName) {
+                it.setupJarManifest(multiRelease = config.jvmVariant.toJvmVariant() == JvmVariant.BOTH)
+            }
         }
         // test should compile and run against original production binaries
         if (compilationType == CompilationType.TEST) {
             val mainCompilation =
                 compilation.target.compilations.getByName(KotlinCompilation.MAIN_COMPILATION_NAME)
-            val originalMainClassesDirs = project.files(
-                // use Callable because there is no guarantee that main is configured before test
-                Callable { originalDirsByCompilation[mainCompilation]!! }
+            val originalMainClassesDirs = project.objects.fileCollection().from(
+                mainCompilation.compileTaskProvider.flatMap { (it as KotlinCompileTool).destinationDirectory }
             )
 
             // KGP >= 1.7.0 has breaking changes in task hierarchy:
```

Comment:
I would propose to remove this if here as now minimal supported KGP version is `1.7.0`